File: sdks/typescript/src/core/rights.ts
/**
 * DCP-08 v2.0 Rights & Obligations.
 *
 * Implements rights declarations, obligation recording, violation reporting,
 * and compliance checking. Rights violations can optionally create disputes
 * via DCP-07 integration.
 */

import type { CompositeKeyPair } from './composite-ops.js';
import { compositeSign } from './composite-ops.js';
import type { AlgorithmRegistry } from './crypto-registry.js';
import { canonicalizeV2 } from './canonicalize.js';
import { DCP_CONTEXTS } from './domain-separation.js';
import type {
  RightType,
  ComplianceStatus,
  RightEntry,
  RightsDeclaration,
  ObligationRecord,
  RightsViolationReport,
} from '../types/v2.js';

/**
 * Declare rights for an agent (typically invoked at commissioning).
 */
export async function declareRights(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    declaration_id: string;
    session_nonce: string;
    agent_id: string;
    rights: RightEntry[];
    jurisdiction: string;
  },
): Promise<RightsDeclaration> {
  const payload = {
    dcp_version: '2.0' as const,
    declaration_id: params.declaration_id,
    session_nonce: params.session_nonce,
    agent_id: params.agent_id,
    rights: params.rights,
    jurisdiction: params.jurisdiction,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Rights, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Record an obligation between an agent and its responsible principal.
 */
export async function recordObligation(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    obligation_id: string;
    session_nonce: string;
    agent_id: string;
    human_id: string;
    obligation_type: string;
    compliance_status: ComplianceStatus;
    evidence_hashes: string[];
  },
): Promise<ObligationRecord> {
  const payload = {
    dcp_version: '2.0' as const,
    obligation_id: params.obligation_id,
    session_nonce: params.session_nonce,
    agent_id: params.agent_id,
    human_id: params.human_id,
    obligation_type: params.obligation_type,
    compliance_status: params.compliance_status,
    evidence_hashes: params.evidence_hashes,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Rights, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Report a rights violation. Optionally links to a DCP-07 dispute.
 */
export async function reportViolation(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    violation_id: string;
    session_nonce: string;
    agent_id: string;
    violated_right: RightType;
    evidence_hashes: string[];
    dispute_id: string | null;
  },
): Promise<RightsViolationReport> {
  const payload = {
    dcp_version: '2.0' as const,
    violation_id: params.violation_id,
    session_nonce: params.session_nonce,
    agent_id: params.agent_id,
    violated_right: params.violated_right,
    evidence_hashes: params.evidence_hashes,
    dispute_id: params.dispute_id,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Rights, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Check rights compliance for an agent against its declared rights.
 */
export function checkRightsCompliance(
  declaration: RightsDeclaration,
  obligations: ObligationRecord[],
): { compliant: boolean; violations: string[] } {
  const violations: string[] = [];

  for (const obligation of obligations) {
    if (obligation.compliance_status === 'non_compliant') {
      violations.push(
        `Obligation ${obligation.obligation_id} (${obligation.obligation_type}) is non-compliant`,
      );
    }
  }

  return {
    compliant: violations.length === 0,
    violations,
  };
}

File: sdks/python/dcp_ai/v2/rights.py
"""
DCP-08 v2.0 Rights & Obligations — Python port.
"""

from __future__ import annotations

from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS
from dcp_ai.v2.lifecycle import _utc_now


async def _sign_rights(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    payload: dict[str, Any],
) -> dict[str, Any]:
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Rights"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


async def declare_rights(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    declaration_id: str,
    session_nonce: str,
    agent_id: str,
    rights: list[dict[str, Any]],
    jurisdiction: str,
) -> dict[str, Any]:
    """Declare rights for an agent (typically invoked at commissioning)."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "declaration_id": declaration_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "rights": rights,
        "jurisdiction": jurisdiction,
        "timestamp": _utc_now(),
    }
    return await _sign_rights(registry, classical_key, pq_key, payload)


async def record_obligation(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    obligation_id: str,
    session_nonce: str,
    agent_id: str,
    human_id: str,
    obligation_type: str,
    compliance_status: str,
    evidence_hashes: list[str],
) -> dict[str, Any]:
    """Record an obligation between an agent and its responsible principal."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "obligation_id": obligation_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "human_id": human_id,
        "obligation_type": obligation_type,
        "compliance_status": compliance_status,
        "evidence_hashes": evidence_hashes,
        "timestamp": _utc_now(),
    }
    return await _sign_rights(registry, classical_key, pq_key, payload)


async def report_violation(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    violation_id: str,
    session_nonce: str,
    agent_id: str,
    violated_right: str,
    evidence_hashes: list[str],
    dispute_id: str | None,
) -> dict[str, Any]:
    """Report a rights violation. Optionally links to a DCP-07 dispute."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "violation_id": violation_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "violated_right": violated_right,
        "evidence_hashes": evidence_hashes,
        "dispute_id": dispute_id,
        "timestamp": _utc_now(),
    }
    return await _sign_rights(registry, classical_key, pq_key, payload)


def check_rights_compliance(
    declaration: dict[str, Any],
    obligations: list[dict[str, Any]],
) -> dict[str, Any]:
    """Check rights compliance. Returns {'compliant': bool, 'violations': [str]}."""
    _ = declaration  # declaration kept in signature for API parity
    violations: list[str] = []
    for obligation in obligations:
        if obligation.get("compliance_status") == "non_compliant":
            violations.append(
                f"Obligation {obligation.get('obligation_id')} "
                f"({obligation.get('obligation_type')}) is non-compliant"
            )
    return {"compliant": not violations, "violations": violations}
